### PR TITLE
First support for creating dependent jobs - Fixes #140, #141, #146

### LINKED ISSUE #140: lookup the latest full job for a 'service.instance' parent
We hide the 'full' name of a chronos job from the user. When a user adds a dependent job, they need to specify a parent as 'service.instance'. When this job is created, we need to do some work to find the 'full' job name.

The acceptance criteria for this is to be able to launch a dependent job where the parent is specified in the service.instance format.

### PR DESCRIPTION
**Pull request by Rob-Johnson**
This is a first pass at adding support for dependent jobs: a user can now add a 'parents' key to the yelpsoa-configs for their job and have a dependent job created.

A user specifies the parent for a dependent job in the format `service.instance`. This value is an array, and must be specified as so in the soa config.

When the job object is formatted into the JSON to be POSTed to Chronos, the 'service.instance' is translated into the 'full' job name by finding the most recent job associated with that service instance.

Caveats:

- The 'parent' job must already be launched in chronos before the dependent job is launched #147
- Validation for a matching job is done at launch time, rather than at 'validate' time.
- parents is an array, and setting the ``parent`` value as a string rather than array will blow up

There is probably more, but these are the ones I can think of immediately.

I've also done some refactoring of the integration tests for chronos as part of this. The key reasons for this were:

- We can now operate on multiple jobs at the same time using the new `jobs` field I added to the context. Before ``context.chronos_job`` was the only key available, which makes it more difficult when I need to operate on multiple jobs at the same time.
- The integration tests for ``setup_chronos_job`` were operating on ChronosJobConfig objects, and the chronos client rather than going through the full workflow the user would go through. I've modified this, so now the features more closely follow the users pattern. E.G 'I have yelpsoa-configs, I run setup_chronos_job, I modify the yelpsoa-configs, I run setup_chronos_job' rather than 'I have a chronosjobconfig, I run setup_chronos_job.setup_job(myconfig), I edit a field in the config etc. I think this is a good thing - the only caveat here is there are exceptions in the itest output from errors connecting to Sensu. Hopefully I can fix that by adding something that acts like a sensu server to listen a discard messages.

I found a bug (#146) as a result of doing this, so I think it was worthwhile. The fix for #146 is also in this PR.

### CODE REVIEW

File: paasta_itests/environment.py [hunk 1]
```diff
@@ -64,6 +64,8 @@ def _clean_up_chronos_jobs(context):
                 print "after_scenario: Job %s is present in chronos. Deleting." % job['name']
                 context.chronos_client.delete(job['name'])
             time.sleep(1)
+    if hasattr(context, 'jobs'):
```
Comment (solarkennedy) [inline paasta_itests/environment.py]: Is this correct, did you want `! hasattr`? If not, what is the purpose of this code?

File: paasta_tools/chronos_tools.py [hunk 2]
```diff
@@ -674,3 +681,23 @@ def delete_job(client, job):
 def create_job(client, job):
     log.debug("Creating job: %s" % job)
     client.add(job)
+
+
+def find_matching_parent_jobs(job_name):
```
Comment (solarkennedy) [inline paasta_tools/chronos_tools.py]: Can you make this singular if this only returns one job?

File: paasta_itests/environment.py (hunk 1 above)
Comment (Rob-Johnson) [inline paasta_itests/environment.py]: hm, this was to make sure that the jobs were sanitised between each scenario.

the 'setup' part of this is done in [here](https://github.com/Yelp/paasta/blob/88defcdedbbd82d4be659d6047cf07fbb5dd88e1/paasta_itests/steps/setup_steps.py#L116). Given it's in the 'given a working paasta cluster', which is done for each scenario (though that's not a formal requirement), it renders the code you've referenced pretty useless. Maybe I should leave this here just to be defensive (in which case, I should move the 'setup' part of it somewhere else.

**Comment by Rob-Johnson** [comment]
huh, I didn't know Travis was smart enough to run tests on the merge of this branch into master, rather than just this branch. looks like there is something in cli/validate that is requiring the ``schedule`` field to exist.

**Comment by solarkennedy** [comment]
I think this is going to be one of those things where the change is so significant that it will be hard to really see the forest through the trees.

Thank you for the itest refactor, it was a long time coming.

We can worry about #147 next, it is the next logical road block, but at least with this code you can manually begin experimenting.

ship!

**Pull request merged**